File: app/src/main/java/com/prolocity/patchtracker/data/PatchAwardLine.kt
package com.prolocity.patchtracker.data

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import java.time.LocalDate

@Entity(
    tableName = "patch_award_lines",
    foreignKeys = [
        ForeignKey(
            entity = PatchAwardEvent::class,
            parentColumns = ["id"],
            childColumns = ["eventId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = PatchType::class,
            parentColumns = ["id"],
            childColumns = ["patchTypeId"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("eventId"), Index("patchTypeId")]
)
data class PatchAwardLine(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val eventId: Long,
    val patchTypeId: Long,
    // True if this specific patch was physically handed to the player when it was earned.
    val awardedAtTime: Boolean,
    // Set once an initially-owed patch is later handed over. Null while still outstanding.
    val fulfilledDate: LocalDate? = null,
    // True if the player chose to enter the Mini Mania raffle instead of taking the physical
    // patch. Mutually exclusive with awardedAtTime/fulfilledDate - a raffle-opted line is never
    // "still owed" since there's no patch handover pending.
    val optedForRaffle: Boolean = false
)

enum class PatchLineStatus { AWARDED, OWED, RAFFLE }

fun patchLineStatus(awardedAtTime: Boolean, fulfilledDate: LocalDate?, optedForRaffle: Boolean): PatchLineStatus = when {
    optedForRaffle -> PatchLineStatus.RAFFLE
    awardedAtTime || fulfilledDate != null -> PatchLineStatus.AWARDED
    else -> PatchLineStatus.OWED
}

val PatchAwardLine.status: PatchLineStatus get() = patchLineStatus(awardedAtTime, fulfilledDate, optedForRaffle)
val PatchAwardLine.isOutstanding: Boolean get() = status == PatchLineStatus.OWED

File: ios/PatchTracker/Data/Models/PatchAwardLine.swift
import Foundation
import SwiftData

/// The three mutually-exclusive states a patch line can be in, derived from its raw fields by
/// `patchLineStatus`. Mirrors the Android `PatchLineStatus` enum.
enum PatchLineStatus: Hashable {
    case awarded, owed, raffle
}

/// Pure derivation shared by the live SwiftData model, the backup wire format, and the
/// display-ready resolved backup — mirrors the Android free function `patchLineStatus`.
func patchLineStatus(awardedAtTime: Bool, fulfilledDate: Date?, optedForRaffle: Bool) -> PatchLineStatus {
    if optedForRaffle { return .raffle }
    if awardedAtTime || fulfilledDate != nil { return .awarded }
    return .owed
}

/// A single patch on an award entry, tracked independently. `awardedAtTime` is true if the
/// patch was physically handed over when earned; `fulfilledDate` is set once an initially-owed
/// patch is later handed over; `optedForRaffle` is true if the player chose to enter the Mini
/// Mania raffle instead of taking the patch — mutually exclusive with the other two at the UI
/// layer. `status`/`isOutstanding` derive the Awarded/Owed/Raffle badge from these three fields.
@Model
final class PatchAwardLine {
    var awardedAtTime: Bool
    var fulfilledDate: Date?
    var optedForRaffle: Bool = false

    var event: PatchAwardEvent?
    var patchType: PatchType?

    init(
        awardedAtTime: Bool,
        fulfilledDate: Date? = nil,
        optedForRaffle: Bool = false,
        event: PatchAwardEvent?,
        patchType: PatchType?
    ) {
        self.awardedAtTime = awardedAtTime
        self.fulfilledDate = fulfilledDate
        self.optedForRaffle = optedForRaffle
        self.event = event
        self.patchType = patchType
    }

    var status: PatchLineStatus {
        patchLineStatus(awardedAtTime: awardedAtTime, fulfilledDate: fulfilledDate, optedForRaffle: optedForRaffle)
    }

    /// Owed and not yet handed over. Mirrors the Android `PatchAwardLine.isOutstanding`. A
    /// raffle-opted line is never outstanding — there's no patch handover pending.
    var isOutstanding: Bool { status == .owed }
}
